pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,172 @@
+	syncUpdatersInterval = time.Minute
+	// updaterWorkerNum represents the maximum number of updaters which can be
+	// run at the same time.
+	maxUpdaterWorkerNum = 10
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	syncUpdatersTicker := time.NewTicker(syncUpdatersInterval)
+	defer syncUpdatersTicker.Stop()
+	syncDeploymentChainsTicker := time.NewTicker(syncDeploymentChainsInterval)
+	defer syncDeploymentChainsTicker.Stop()
+
+	d.logger.Info("start running deployment chain controller")
+	defer d.logger.Info("deployment chain controller has been stopped")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-syncUpdatersTicker.C:
+			if err := d.syncUpdaters(ctx); err != nil {
+				d.logger.Error("failed while sync controller updaters", zap.Error(err))
+			}
+		case <-syncDeploymentChainsTicker.C:
+			if err := d.syncDeploymentChains(ctx); err != nil {
+				d.logger.Error("failed while sync deployment chains", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (d *DeploymentChainController) syncUpdaters(ctx context.Context) error {
+	// Remove done updater of completed deployment chain.
+	for id, u := range d.updaters {
+		if u.IsDone() {
+			d.logger.Info("remove done updater of deployment chain",
+				zap.String("id", id),
+				zap.Time("completed_at", u.doneTimestamp),
+			)
+			delete(d.updaters, id)
+		}
+	}
+
+	// Find all not completed deployment chains and create updater if does not exist.
+	notCompletedChains, err := listNotCompletedDeploymentChain(ctx, d.deploymentChainStore)
+	if err != nil {
+		d.logger.Error("failed to fetch all not completed deployment chain", zap.Error(err))
+		return err
+	}
+	for _, c := range notCompletedChains {
+		if _, ok := d.updaters[c.Id]; !ok {
+			d.updaters[c.Id] = newUpdater(
+				c,
+				d.applicationStore,
+				d.deploymentStore,
+				d.deploymentChainStore,
+				d.logger,
+			)
+		}
+	}
+
+	return nil
+}
+
+func (d *DeploymentChainController) syncDeploymentChains(ctx context.Context) error {
+	var (
+		dcUpdatersNum = len(d.updaters)
+		updatersCh    = make(chan *updater, dcUpdatersNum)
+		resultCh      = make(chan error, dcUpdatersNum)
```

Comment:
nit: We can just use a `WaitGroup` to wait for their completions since the returned error doesn't matter.